In [31]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
dataset1=pd.read_csv('C:/Users/RUSHI/Downloads/general_data.csv') 
dataset1.head(15)

Age Attrition     BusinessTravel              Department  \
0    51        No      Travel_Rarely                   Sales   
1    31       Yes  Travel_Frequently  Research & Development   
2    32        No  Travel_Frequently  Research & Development   
3    38        No         Non-Travel  Research & Development   
4    32        No      Travel_Rarely  Research & Development   
5    46        No      Travel_Rarely  Research & Development   
6    28       Yes      Travel_Rarely  Research & Development   
7    29        No      Travel_Rarely  Research & Development   
8    31        No      Travel_Rarely  Research & Development   
9    25        No         Non-Travel  Research & Development   
10   45        No      Travel_Rarely  Research & Development   
11   36        No      Travel_Rarely  Research & Development   
12   55        No      Travel_Rarely  Research & Development   
13   47       Yes         Non-Travel  Research & Development   
14   28        No      Travel_Rarely  Research & Development   

    DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                  6          2  Life Sciences              1           1   
1                 10          1  Life Sciences              1           2   
2                 17          4          Other              1           3   
3                  2          5  Life Sciences              1           4   
4                 10          1        Medical              1           5   
5                  8          3  Life Sciences              1           6   
6                 11          2        Medical              1           7   
7                 18          3  Life Sciences              1           8   
8                  1          3  Life Sciences              1           9   
9                  7          4        Medical              1          10   
10                17          2        Medical              1          11   
11                28          1  Life Sciences              1          12   
12                14          4  Life Sciences              1          13   
13                 1          1        Medical              1          14   
14                 1          3  Life Sciences              1          15   

    Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0   Female  ...                 1.0      Y                11              8   
1   Female  ...                 0.0      Y                23              8   
2     Male  ...                 1.0      Y                15              8   
3     Male  ...                 3.0      Y                11              8   
4     Male  ...                 4.0      Y                12              8   
5   Female  ...                 3.0      Y                13              8   
6     Male  ...                 2.0      Y                20              8   
7     Male  ...                 2.0      Y                22              8   
8     Male  ...                 0.0      Y                21              8   
9   Female  ...                 1.0      Y                13              8   
10    Male  ...                 0.0      Y                13              8   
11    Male  ...                 0.0      Y                12              8   
12  Female  ...                 0.0      Y                17              8   
13    Male  ...                 1.0      Y                11              8   
14    Male  ...                 1.0      Y                14              8   

    StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                  0               1.0                      6               1   
1                  1               6.0                      3               5   
2                  3               5.0                      2               5   
3                  3              13.0                      5               8   
4                  2               9.0                      2               

In [35]:
dataset1.isnull()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  \
0     False      False           False       False             False   
1     False      False           False       False             False   
2     False      False           False       False             False   
3     False      False           False       False             False   
4     False      False           False       False             False   
...     ...        ...             ...         ...               ...   
4405  False      False           False       False             False   
4406  False      False           False       False             False   
4407  False      False           False       False             False   
4408  False      False           False       False             False   
4409  False      False           False       False             False   

      Education  EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0         False           False          False       False   False  ...   
1         False           False          False       False   False  ...   
2         False           False          False       False   False  ...   
3         False           False          False       False   False  ...   
4         False           False          False       False   False  ...   
...         ...             ...            ...         ...     ...  ...   
4405      False           False          False       False   False  ...   
4406      False           False          False       False   False  ...   
4407      False           False          False       False   False  ...   
4408      False           False          False       False   False  ...   
4409      False           False          False       False   False  ...   

      NumCompaniesWorked  Over18  PercentSalaryHike  StandardHours  \
0                  False   False              False          False   
1                  False   False              False          False   
2                  False   False              False          False   
3                  False   False              False          False   
4                  False   False              False          False   
...                  ...     ...                ...            ...   
4405               False   False              False          False   
4406               False   False              False          False   
4407               False   False              False          False   
4408               False   False              False          False   
4409               False   False              False          False   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                False              False                  False   
1                False              False                  False   
2                False              False                  False   
3                False              False                  False   
4                False              False                  False   
...                ...                ...                    ...   
4405             False              False                  False   
4406             False              False                  False   
4407             False              False                  False   
4408             False              False                  False   
4409             False               True                  False   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0              False                    False                 False  
1              False                    False                 False  
2              False                    False                 False  
3              False                    False                 False  
4              False                    False                 False  
...              ...                      ...                   ...  
4405           False                    False                 False  
4406           False    

In [40]:
dataset1.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4405    False
4406    False
4407    False
4408    False
4409    False
Length: 4410, dtype: bool

In [41]:
dataset1.drop_duplicates()

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ..

In [43]:
dataset1.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [44]:
dataset1.describe()

Age  DistanceFromHome    Education  EmployeeCount   EmployeeID  \
count  4410.000000       4410.000000  4410.000000         4410.0  4410.000000   
mean     36.923810          9.192517     2.912925            1.0  2205.500000   
std       9.133301          8.105026     1.023933            0.0  1273.201673   
min      18.000000          1.000000     1.000000            1.0     1.000000   
25%      30.000000          2.000000     2.000000            1.0  1103.250000   
50%      36.000000          7.000000     3.000000            1.0  2205.500000   
75%      43.000000         14.000000     4.000000            1.0  3307.750000   
max      60.000000         29.000000     5.000000            1.0  4410.000000   

          JobLevel  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
count  4410.000000    4410.000000         4391.000000        4410.000000   
mean      2.063946   65029.312925            2.694830          15.209524   
std       1.106689   47068.888559            2.498887           3.659108   
min       1.000000   10090.000000            0.000000          11.000000   
25%       1.000000   29110.000000            1.000000          12.000000   
50%       2.000000   49190.000000            2.000000          14.000000   
75%       3.000000   83800.000000            4.000000          18.000000   
max       5.000000  199990.000000            9.000000          25.000000   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count         4410.0       4410.000000        4401.000000   
mean             8.0          0.793878          11.279936   
std              0.0          0.851883           7.782222   
min              8.0          0.000000           0.000000   
25%              8.0          0.000000           6.000000   
50%              8.0          1.000000          10.000000   
75%              8.0          1.000000          15.000000   
max              8.0          3.000000          40.000000   

       TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
count            4410.000000     4410.000000              4410.000000   
mean                2.799320        7.008163                 2.187755   
std                 1.288978        6.125135                 3.221699   
min                 0.000000        0.000000                 0.000000   
25%                 2.000000        3.000000                 0.000000   
50%                 3.000000        5.000000                 1.000000   
75%                 3.000000        9.000000                 3.000000   
max                 6.000000       40.000000                15.000000   

       YearsWithCurrManager  
count           4410.000000  
mean               4.123129  
std                3.567327  
min                0.000000  
25%                2.000000  
50%                3.000000  
75%                7.000000  
max               17.000000

In [45]:
dataset3=dataset1[['Age','DistanceFromHome','Education','MonthlyIncome', 'NumCompaniesWorked', 'PercentSalaryHike',
                   'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany','YearsSinceLastPromotion', 'YearsWithCurrManager']].describe() 
dataset3

Age  DistanceFromHome    Education  MonthlyIncome  \
count  4410.000000       4410.000000  4410.000000    4410.000000   
mean     36.923810          9.192517     2.912925   65029.312925   
std       9.133301          8.105026     1.023933   47068.888559   
min      18.000000          1.000000     1.000000   10090.000000   
25%      30.000000          2.000000     2.000000   29110.000000   
50%      36.000000          7.000000     3.000000   49190.000000   
75%      43.000000         14.000000     4.000000   83800.000000   
max      60.000000         29.000000     5.000000  199990.000000   

       NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  \
count         4391.000000        4410.000000        4401.000000   
mean             2.694830          15.209524          11.279936   
std              2.498887           3.659108           7.782222   
min              0.000000          11.000000           0.000000   
25%              1.000000          12.000000           6.000000   
50%              2.000000          14.000000          10.000000   
75%              4.000000          18.000000          15.000000   
max              9.000000          25.000000          40.000000   

       TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
count            4410.000000     4410.000000              4410.000000   
mean                2.799320        7.008163                 2.187755   
std                 1.288978        6.125135                 3.221699   
min                 0.000000        0.000000                 0.000000   
25%                 2.000000        3.000000                 0.000000   
50%                 3.000000        5.000000                 1.000000   
75%                 3.000000        9.000000                 3.000000   
max                 6.000000       40.000000                15.000000   

       YearsWithCurrManager  
count           4410.000000  
mean               4.123129  
std                3.567327  
min                0.000000  
25%                2.000000  
50%                3.000000  
75%                7.000000  
max               17.000000

In [48]:
dataset3=dataset1[['Age','DistanceFromHome','Education','MonthlyIncome', 'NumCompaniesWorked', 'PercentSalaryHike','TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany','YearsSinceLastPromotion', 'YearsWithCurrManager']].median()

In [49]:
dataset3


Age                           36.0
DistanceFromHome               7.0
Education                      3.0
MonthlyIncome              49190.0
NumCompaniesWorked             2.0
PercentSalaryHike             14.0
TotalWorkingYears             10.0
TrainingTimesLastYear          3.0
YearsAtCompany                 5.0
YearsSinceLastPromotion        1.0
YearsWithCurrManager           3.0
dtype: float64

In [50]:
dataset3=dataset1[['Age','DistanceFromHome','Education','MonthlyIncome', 'NumCompaniesWorked', 'PercentSalaryHike',
                   'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany','YearsSinceLastPromotion',
                   'YearsWithCurrManager']].mode() 
dataset3

Age  DistanceFromHome  Education  MonthlyIncome  NumCompaniesWorked  \
0   35                 2          3          23420                 1.0   

   PercentSalaryHike  TotalWorkingYears  TrainingTimesLastYear  \
0                 11               10.0                      2   

   YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0               5                        0                     2

In [52]:
dataset3=dataset1[['Age','DistanceFromHome','Education','MonthlyIncome',
                   'NumCompaniesWorked', 'PercentSalaryHike','TotalWorkingYears',
                   'TrainingTimesLastYear', 'YearsAtCompany','YearsSinceLastPromotion', 'YearsWithCurrManager']].var() 
dataset3

Age                        8.341719e+01
DistanceFromHome           6.569144e+01
Education                  1.048438e+00
MonthlyIncome              2.215480e+09
NumCompaniesWorked         6.244436e+00
PercentSalaryHike          1.338907e+01
TotalWorkingYears          6.056298e+01
TrainingTimesLastYear      1.661465e+00
YearsAtCompany             3.751728e+01
YearsSinceLastPromotion    1.037935e+01
YearsWithCurrManager       1.272582e+01
dtype: float64

In [53]:
dataset3=dataset1[['Age','DistanceFromHome','Education',
                   'MonthlyIncome', 'NumCompaniesWorked', 'PercentSalaryHike','TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany',
                   'YearsSinceLastPromotion', 'YearsWithCurrManager']].skew()
dataset3

Age                        0.413005
DistanceFromHome           0.957466
Education                 -0.289484
MonthlyIncome              1.368884
NumCompaniesWorked         1.026767
PercentSalaryHike          0.820569
TotalWorkingYears          1.116832
TrainingTimesLastYear      0.552748
YearsAtCompany             1.763328
YearsSinceLastPromotion    1.982939
YearsWithCurrManager       0.832884
dtype: float64

In [54]:
dataset3=dataset1[['Age','DistanceFromHome','Education',
                   'MonthlyIncome', 'NumCompaniesWorked', 'PercentSalaryHike','TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany',
                   'YearsSinceLastPromotion', 'YearsWithCurrManager']].kurt()
dataset3

Age                       -0.405951
DistanceFromHome          -0.227045
Education                 -0.560569
MonthlyIncome              1.000232
NumCompaniesWorked         0.007287
PercentSalaryHike         -0.302638
TotalWorkingYears          0.912936
TrainingTimesLastYear      0.491149
YearsAtCompany             3.923864
YearsSinceLastPromotion    3.601761
YearsWithCurrManager       0.167949
dtype: float64

Age                       -0.405951
DistanceFromHome          -0.227045
Education                 -0.560569
MonthlyIncome              1.000232
NumCompaniesWorked         0.007287
PercentSalaryHike         -0.302638
TotalWorkingYears          0.912936
TrainingTimesLastYear      0.491149
YearsAtCompany             3.923864
YearsSinceLastPromotion    3.601761
YearsWithCurrManager       0.167949
dtype: float64
Inference from the analysis:

#### All the above variables show positive skewness; while Age & Mean_distance_from_home are leptokurtic and all other variables are platykurtic.

#### The Mean_Monthly_Income’s IQR is at 54K suggesting company wide attrition across all income bands

#### Mean age forms a near normal distribution with 13 years of IQR

In [ ]:

Goal of the case study You are required to model the probability of attrition. The results thus obtained will be used by the management to understand what changes they should make to their workplace, in order to get most of their employees to stay.

Columns
EmployeeIDEmployee number/id
EnvironmentSatisfactionWork Environment Satisfaction Level
JobSatisfactionJob Involvement Level Job Involvement Level Job Involvement Level
WorkLifeBalanceWork life balance level